pkg/reacher/reacher.go

```diff
@@ -0,0 +1,114 @@
+package reacher
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/ethersphere/bee/pkg/p2p"
+	"github.com/ethersphere/bee/pkg/swarm"
+	ma "github.com/multiformats/go-multiaddr"
+)
+
+const pingTimeout = time.Second * 5
+
+type peer struct {
+	overlay swarm.Address
+	addr    ma.Multiaddr
+}
+
+type reacher struct {
+	mu    sync.Mutex
+	queue []peer
+
+	quit chan struct{}
+	run  chan struct{}
+
+	streamer p2p.StreamerPinger
+	notifier p2p.ReachableNotifier
+
+	wg sync.WaitGroup
+}
+
+func New(streamer p2p.StreamerPinger, notifier p2p.ReachableNotifier) *reacher {
+
+	r := &reacher{
+		quit:     make(chan struct{}),
+		run:      make(chan struct{}, 1),
+		streamer: streamer,
+		notifier: notifier,
+		wg:       sync.WaitGroup{},
+	}
+
+	r.wg.Add(1)
+	go r.worker()
+
+	return r
+}
+
+func (r *reacher) worker() {
```

Comment:
the worker and `ping` can be condensed into one method potentially, where you can just start an extra goroutine that cancels the context when `r.quit` is closed. also, limiting the number of peers we try with a semaphore would be nice. another thing is that it would be nice to have multiple retries with some exponential backoff and some reasonable upper bound to limit how long it should take to resolve the status of the peer